maven-extension/src/main/java/io/opentelemetry/maven/SpanRegistry.java

```diff
@@ -0,0 +1,183 @@
+    if (this.rootSpan != null) {
+      throw new IllegalStateException("Root span already defined " + this.rootSpan);
+    }
+    this.rootSpan = rootSpan;
+  }
+
+  public Span getSpan(@Nonnull MavenProject mavenProject) {
+    final MavenProjectKey key = MavenProjectKey.builder().fromMavenProject(mavenProject).build();
+    final Span span = this.mavenProjectKeySpanMap.get(key);
+    if (span == null) {
+      throw new IllegalStateException(
+          "Span not started for project "
+              + mavenProject.getGroupId()
+              + ":"
+              + mavenProject.getArtifactId());
+    }
+    return span;
+  }
+
+  public Span getRootSpanNotNull() {
+    if (rootSpan == null) {
+      throw new IllegalStateException("Root span not defined");
+    }
+    return rootSpan;
+  }
+
+  public Span removeRootSpan() {
+    if (rootSpan == null) {
+      throw new IllegalStateException("Root span not defined");
+    }
+    if (!this.mojoExecutionKeySpanMap.isEmpty()) {
+      throw new IllegalStateException(
+          "Remaining children spans: "
+              + this.mojoExecutionKeySpanMap.keySet().stream()
+                  .map(MojoExecutionKey::toString)
+                  .collect(Collectors.joining(", ")));
+    }
+    return rootSpan;
+  }
+
+  public void putSpan(Span span, MavenProject mavenProject) {
+    MavenProjectKey key = MavenProjectKey.builder().fromMavenProject(mavenProject).build();
+    Span previousSpanForKey = mavenProjectKeySpanMap.put(key, span);
+    if (previousSpanForKey != null) {
+      throw new IllegalStateException("A span has already been started for " + mavenProject);
+    }
+  }
+
+  public Span removeSpan(MavenProject mavenProject) throws IllegalStateException {
+    MavenProjectKey key = MavenProjectKey.builder().fromMavenProject(mavenProject).build();
+    Span span = mavenProjectKeySpanMap.remove(key);
+    if (span == null) {
+      throw new IllegalStateException("No span found for " + mavenProject);
+    }
+    return span;
+  }
+
+  public void putSpan(Span span, MojoExecution mojoExecution) {
+    MojoExecutionKey key = MojoExecutionKey.builder().fromMojoExecution(mojoExecution).build();
+    Span previousSpanForKey = mojoExecutionKeySpanMap.put(key, span);
+    if (previousSpanForKey != null) {
+      throw new IllegalStateException("A span has already been started for " + mojoExecution);
+    }
+  }
+
+  @Nonnull
+  public Span removeSpan(MojoExecution mojoExecution) throws IllegalStateException {
+    MojoExecutionKey key = MojoExecutionKey.builder().fromMojoExecution(mojoExecution).build();
+    Span span = mojoExecutionKeySpanMap.remove(key);
+    if (span == null) {
+      throw new IllegalStateException("No span found for " + mojoExecution);
+    }
+    return span;
+  }
+
+  @AutoValue
+  abstract static class MavenProjectKey {
+    abstract String groupId();
+
+    abstract String artifactId();
+
+    public static Builder builder() {
```

Comment:
It looks like my example may have caused confusion - the usages have the `from` pattern so I think the Builder is unnecessary. You can remove the Builder class, which will allow calling the autovalue constructor directly instead

```java
public static MavenProjectKey.fromMavenProject(MavenProject mavenProject)) {
  return new AutoValue_SpanRegistry_MavenProjectKey(mavenProject.getGroupId(), mavenProject.getArtifactId());
}
```